### PR TITLE
[master] Cannot set FOR UPDATE in conjunction with ORDER BY and / or LIMIT (OracleDB & Derby) - bugfix + unit test

### LINKED ISSUE #1369: Cannot set FOR UPDATE in conjunction with ORDER BY and / or LIMIT (OracleDB & Derby)
Running against OracleDB or Derby, if ORDER BY is set on CriteriaQuery, and / or setMaxResults is called on TypedQuery object, then a lock instruction applied using

```
final TypedQuery<Long> query = em.createQuery(cq);
query.setLockMode(LockModeType.PESSIMISTIC_WRITE);
```
results in syntax error on OracleDB, while against Derby the instruction appears to be ignored. Running against SQLServer this works as expected.

Using `query.unwrap(JpaQuery.class).getDatabaseQuery().getSQLString()` to obtain the resultant SQL statement.

Running the following:
```
            cq.orderBy( cb.asc( jpaTable.get(JPATABLE_.getId()) ) );
            final TypedQuery<Long> query = em.createQuery(cq);
            query.setMaxResults(pageSize);
            query.setLockMode(LockModeType.PESSIMISTIC_WRITE);
            final List<Long> ids = query.getResultList();
```
Although the row limit doesn't appear to be applied in the SQL statement, it is enforced somehow, because we do get correct result: `SELECT ID FROM JPATABLE WITH (UPDLOCK) WHERE (TENANT = ?) ORDER BY ID ASC`

Against Derby, the FOR UPDATE instruction seems to be ignored - `SELECT ID AS a1 FROM JPATABLE WHERE (TENANT = ?) ORDER BY ID ASC OFFSET ? ROWS FETCH NEXT ? ROWS ONLY`

Against OracleDB, this results in an exception:
    _javax.persistence.PessimisticLockException: Exception [EclipseLink-4002] (Eclipse Persistence Services - 2.7.8.v20201217-ecdf3c32c4): org.eclipse.persistence.exceptions.DatabaseException
    Internal Exception: java.sql.SQLSyntaxErrorException: ORA-00933: SQL command not properly ended
    Error Code: 933_

Running a query with row limit and ORDER BY, but with no locking specified works correct against all DB types. Running a query with locking specified, and without ORDER BY and row limit also works fine.

### PR DESCRIPTION
**Pull request by rfelcman**
Fixes #1369 .

Signed-off-by: Radek Felcman <radek.felcman@oracle.com>

### CODE REVIEW

**Comment by rfelcman** [comment]
There are tests results CORE, JPA, JPASE against Oracle 12c.
[test-reports.tar.gz](https://github.com/eclipse-ee4j/eclipselink/files/7861837/test-reports.tar.gz)

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/SQLSelectStatement.java [hunk 1]
```diff
@@ -1722,69 +1722,102 @@ protected void normalizeOrderBy(Expression builder, List<Expression> allExpressi
     /**
      * Print the SQL representation of the statement on a stream.
      */
-    public Vector<DatabaseField> printSQL(ExpressionSQLPrinter printer) {
+    public Vector printSQL(ExpressionSQLPrinter printer) {
```
Comment (lukasj) [inline foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/SQLSelectStatement.java:1725]: why generics were removed?

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/platform/database/OraclePlatform.java [hunk 2]
```diff
@@ -951,14 +951,16 @@ public boolean useJDBCStoredProcedureSyntax() {
     protected String MAX_ROW = "WHERE ROWNUM <= ";
     protected String MIN_ROW = ") WHERE rnum > ";
     // Bug #453208
-    protected String LOCK_START_PREFIX = " AND (";
+    protected String LOCK_START_PREFIX_AND = " AND (";
```
Comment (lukasj) [inline foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/platform/database/OraclePlatform.java:954]: is this rename required? Keep in mind that during the backport, one will have to keep the original name (compatibility)

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/SQLSelectStatement.java [hunk 3]
```diff
@@ -1722,69 +1722,102 @@ protected void normalizeOrderBy(Expression builder, List<Expression> allExpressi
...
 
-            if (hasGroupByExpressions()) {
-                appendGroupByClauseToWriter(printer);
-            }
-            if (hasHavingExpression()) {
-                //appendHavingClauseToWriter(printer);
-                printer.printString(" HAVING ");
-                printer.printExpression(getHavingExpression());
-            }
+    public void printSQLWhereKeyWord(ExpressionSQLPrinter printer) throws IOException {
```
Comment (lukasj) [inline foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/SQLSelectStatement.java:1773]: do these new methods have to be public instead of private/protected?

Comment (rfelcman) [reply foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/SQLSelectStatement.java:1773]: Yes because they are called from `org.eclipse.persistence.platform.database.OraclePlatform` (different package).

**Review by lukasj — APPROVED** [review]
LGTM

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/SQLSelectStatement.java (hunk 3 above)
Comment (lukasj) [reply foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/SQLSelectStatement.java:1773]: ok
